File: data_structures/Graph/Graph.js
class Graph {
  constructor() {
    this.adjacencyList = {};
  }

  addVertex = (vertex) => {
    if (!this.adjacencyList[vertex]) {
      this.adjacencyList[vertex] = [];
    } else {
      console.log("Vertex aleady exists");
    }
  };

  addEdge = (vertex1, vertex2) => {
    if (this.adjacencyList[vertex1] && this.adjacencyList[vertex2]) {
      if (!this.adjacencyList[vertex1].find((vertex) => vertex == vertex2)) {
        this.adjacencyList[vertex1].push(vertex2);
        this.adjacencyList[vertex2].push(vertex1);
      } else {
        console.log("Edge already exists");
      }
    } else {
      console.log("cannot add edge when vertex does not exist");
    }
  };

  removeEdge = (vertex1, vertex2) => {
    if (this.adjacencyList[vertex1] && this.adjacencyList[vertex2]) {
      this.adjacencyList[vertex1] = this.adjacencyList[vertex1].filter(
        (item) => item !== vertex2
      );
      this.adjacencyList[vertex2] = this.adjacencyList[vertex2].filter(
        (item) => item !== vertex1
      );
    } else {
      console.log("Cannot remove an edge that does not exist");
    }
  };

  removeVertex = (vertex) => {
    if (this.adjacencyList[vertex]) {
      this.adjacencyList[vertex].map((item) => this.removeEdge(vertex, item));
      delete this.adjacencyList[vertex];
    } else {
      console.log("Vertex does not exist");
    }
  };
}

socialGraph = new Graph();
socialGraph.addVertex("Ernest");
socialGraph.addVertex("Obama");
socialGraph.addVertex("Arianna");
socialGraph.addVertex("Betty");
socialGraph.addVertex("Bernard");
socialGraph.addVertex("Josephine");

socialGraph.addEdge("Ernest", "Obama");
socialGraph.addEdge("Ernest", "Arianna");
socialGraph.addEdge("Obama", "Betty");
socialGraph.addEdge("Ernest", "Bernard");
socialGraph.addEdge("Ernest", "Josephine");
socialGraph.addEdge("Ernest", "Betty");
socialGraph.addEdge("Bernard", "Betty");
socialGraph.addEdge("Ernest", "Bernard");
console.log(socialGraph.adjacencyList);

socialGraph.removeEdge("Ernest", "Obama");
socialGraph.removeEdge("Ernest", "Betty");
socialGraph.removeEdge("Ernest", "Arianna");
console.log(socialGraph.adjacencyList);

socialGraph.removeVertex("Ernest");
console.log(socialGraph.adjacencyList);

File: data_structures/Graph/Graph.py
# Author: Ernest Wambua
# Date: 28/12/20
# Purpose: Undirected graph structure


class Graph:
    def __init__(self):
        self.adjacencyList = {}

    def addVertex(self, vertex):
        '''
        Adding a vertex to our graph structure
        '''
        if not self.adjacencyList.get(vertex):
            self.adjacencyList[vertex] = []

    def addEdge(self, vertex1, vertex2):
        if vertex1 in self.adjacencyList.keys() and vertex2 in self.adjacencyList.keys():
            if vertex1 not in self.adjacencyList[vertex2]:
                self.adjacencyList[vertex1].append(vertex2)
                self.adjacencyList[vertex2].append(vertex1)

    def removeEdge(self, vertex1, vertex2):
        if vertex1 in self.adjacencyList.keys() and vertex2 in self.adjacencyList.keys():
            if vertex1 in self.adjacencyList[vertex2]:
                self.adjacencyList[vertex1].remove(vertex2)
                self.adjacencyList[vertex2].remove(vertex1)

    def removeVertex(self, vertex):
        if vertex in self.adjacencyList.keys():
            while len(self.adjacencyList[vertex]) != 0:
                self.removeEdge(vertex, self.adjacencyList[vertex][0])
            self.adjacencyList.pop(vertex)


def main():
    socialGraph = Graph()
    socialGraph.addVertex("Ernest")
    socialGraph.addVertex("Obama")
    socialGraph.addVertex("Arianna")
    socialGraph.addVertex("Betty")
    socialGraph.addVertex("Bernard")
    socialGraph.addVertex("Josephine")

    print(socialGraph.adjacencyList)

    socialGraph.addEdge("Ernest", "Obama")
    socialGraph.addEdge("Ernest", "Arianna")
    socialGraph.addEdge("Obama", "Betty")
    socialGraph.addEdge("Ernest", "Bernard")
    socialGraph.addEdge("Ernest", "Josephine")
    socialGraph.addEdge("Ernest", "Betty")
    socialGraph.addEdge("Bernard", "Betty")
    socialGraph.addEdge("Ernest", "Bernard")

    print(socialGraph.adjacencyList)

    socialGraph.removeEdge("Ernest", "Obama")
    socialGraph.removeEdge("Ernest", "Betty")
    socialGraph.removeEdge("Ernest", "Arianna")

    print(socialGraph.adjacencyList)

    socialGraph.removeVertex("Ernest")

    print(socialGraph.adjacencyList)


if __name__ == "__main__":
    main()
